src/utils/futures_util/src/misc.rs

```diff
@@ -76,3 +77,39 @@ pub async fn await_future_with_monitor_error_stream<T, E, F: Future>(
         Either::Right((output, _)) => Ok(output),
     }
 }
+
+pub struct AttachedFuture<F, T> {
```

Comment:
Don't quite get this, can we add some docstring on its purpose?

Reply:
I have adde the docstring. 

In short, `AttachedFuture::new(future, item)` is similar to `future.map(|output| (output, item))`. The only difference is that the `AttachedFuture` provides method `into_inner` to return the inner future. In practice, we will wrap the join handle with `AttachedFuture` as a future to pass to `FuturesUnordered`. In recovery, we will drain all futures and take the inner join handle to call abort on it to abort and join the future.